### PR TITLE
Add support for toolchain java 15

### PR DESCRIPTION
**Pull request by Jonpez2**
Closes https://github.com/bazelbuild/bazel/issues/11871

I modelled this on @davido's addition of a java 14 toolchain, here:[https://github.com/bazelbuild/bazel/pull/11514](url).

The test plan has the same form: 

Create java_tools from this commit:
`  $ bazel build src:java_tools_java11.zip`

Switch to using the java_tools from the above step in WORKSPACE file:

```
  load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
  http_archive(
    name = "remote_java_tools_linux",
    urls = [
        "file:///<path to the java_tools_java11.zip>",
    ],
  )

```
Add Zulu OpenJDK15 to use as javabase in WORKSPACE file:
```
  http_archive(
    name = "openjdk15_linux_archive",
    build_file_content = """
java_runtime(name = 'runtime', srcs =  glob(['**']), visibility = ['//visibility:public'])
exports_files(["WORKSPACE"], visibility = ["//visibility:public"])
""",
    strip_prefix = "zulu15.27.17-ca-jdk15.0.0-linux_x64",
    urls = ["https://cdn.azul.com/zulu/bin/zulu15.27.17-ca-jdk15.0.0-linux_x64.tar.gz"],
  )
```

Activate custom java toolchain and javabase in .bazelrc file:
```
  build --java_toolchain=@remote_java_tools_linux//:toolchain_jdk_15
  build --host_java_toolchain=@remote_java_tools_linux//:toolchain_jdk_15
  build --javabase=@openjdk15_linux_archive//:runtime
  build --host_javabase=@openjdk15_linux_archive//:runtime

```
Create Java 15 example class file:
```
public class Javac15Example {
  static String textBlock = """
              Hello,
              World
              """;
  public static void main(String[] args) {
    System.out.println(textBlock);
  }
}

```
Add Bazel file to build Java 15 syntax class:
```
  java_binary(
    name = "Javac15Example",
    srcs = ["Javac15Example.java"],
    main_class = "Javac15Example",
  )

```
Test that it works as expected:
```
  $ bazel run java:Javac15Example
INFO: Analyzed target //java:Javac15Example (1 packages loaded, 2 targets configured).
INFO: Found 1 target...
INFO: From Building java/Javac15Example.jar (1 source file):
Target //java:Javac15Example up-to-date:
  bazel-bin/java/Javac15Example.jar
  bazel-bin/java/Javac15Example
INFO: Elapsed time: 1.502s, Critical Path: 1.30s
INFO: 1 process: 1 worker.
INFO: Build completed successfully, 2 total actions
INFO: Build completed successfully, 2 total actions
Hello,
World
```

### CODE REVIEW

**Comment by Jonpez2** [comment]
@googlebot I signed it!

File: WORKSPACE [hunk 1]
```diff
@@ -81,7 +81,7 @@ list_source_repository(name = "local_bazel_source_list")
 # android_sdk_repository(name = "androidsdk")
 # android_ndk_repository(name = "androidndk")
 
-# In order to run //src/test/shell/bazel:maven_skylark_test, follow the
+# In order to run //src/test/shell/bazel:maven_starlark_test, follow the
```
Comment (davido) [inline WORKSPACE:84]: I see some unrelated white space changes. I would rather revert them all, and only include related addition related to java_toolchain_jdk15. Clean up changes should be done in their own PR. Just guessing: you have run buildifier on this files, that also changed the white spaces and shuffled around the attributes.

File: tools/jdk/BUILD.java_tools [hunk 2]
```diff
@@ -158,6 +158,45 @@ java_toolchain(
     target_version = "14",
 )
 
+# A toolchain that targets java 15.
+java_toolchain(
+    name = "toolchain_jdk_15",
```
Comment (davido) [inline tools/jdk/BUILD.java_tools:163]: This could be combined in one single java_toolchain definition, together with "toolchain_jdk_14". Note that this was already the case for "toolchain_jdk_11" and "toolchain_jdk_12", in previous version of this file, but "toolchain_jdk_12" was  removed recently, and "toolchain_jdk_14 was very different compared to toolchain_jdk_11, so that reviewers suggested to split "toolchain_jdk_14" in own rule. That can be re-joined again with "toolchain_jdk_15".

**Review by davido — CHANGES_REQUESTED** [review]
Thanks for the PR. Can you check why CI is failing: [1]? The error is:

(09:53:10) ERROR: /workdir/src/BUILD:744:10: no such package '@remote_java_tools_javac15_test_windows//': The repository '@remote_java_tools_javac15_test_windows' could not be resolved and referenced by '//src:test_repos'.

https://buildkite.com/bazel/bazel-bazel-github-presubmit/builds/7032#fa4a0465-296c-4186-a234-6a72643b54d1

**Comment by davido** [comment]
@Jonpez2 Thanks for the PR.

Could you please reference this https://github.com/bazelbuild/bazel/issues/11871 in the commit message?

//Cc @lberki @comius @philwo @meteorcloudy 

Before this PR is merged we should upload Zulu JDK15 archives to bazel mirror.

File: tools/jdk/BUILD.java_tools (hunk 2 above)
Comment (Jonpez2) [reply tools/jdk/BUILD.java_tools:163]: I don't think I can do this, because of the source_version/target_version attributes.  I could do a list comprehension though if that would make life cleaner?

File: WORKSPACE (hunk 1 above)
Comment (Jonpez2) [reply WORKSPACE]: Done

**Comment by Jonpez2** [comment]
I fixed the specific error you highlighted, @davido , but I'm quite lost on the current set - I can reproduce the failure locally but the error message gives me no hints...

File: src/upload_all_java_tools.sh [hunk 3]
```diff
@@ -48,7 +48,7 @@ bazel_version=$(bazel info release | cut -d' ' -f2)
 
 # Passing the same commit_hash and timestamp to all targets to mark all the artifacts
 # uploaded on GCS with the same identifier.
-for java_version in 11 14; do
+for java_version in 11 14 15; do
```
Comment (comius) [inline src/upload_all_java_tools.sh:51]: Can you rebase to master and leave only 11 here. With a recent commit to master I realised we can have only a single release of java_toolchains, no need to multiply here.

File: tools/jdk/BUILD.java_tools (hunk 2 above)
Comment (comius) [reply tools/jdk/BUILD.java_tools:163]: davido@ This was possible because of multiple releases of Java tools and substitution of JAVA_LANGUAGE_LEVEL. To ease releases of java_tools I prefer to have one version and toolchain_jdk_15 duplicated as it is.

There are other changes coming to make Java use toolchain resolution, where this will probably be cleaned up.

**Review by comius — CHANGES_REQUESTED** [review]
Thanks @Jonpez2 for your effort! It seems that for one time we will catch up with Java releases.

**Comment by comius** [comment]
I uploaded zulu15 to bazel mirror.

File: src/upload_all_java_tools.sh (hunk 3 above)
Comment (Jonpez2) [reply src/upload_all_java_tools.sh]: Done.

File: src/BUILD [hunk 4]
```diff
@@ -570,7 +570,7 @@ JAVA_TOOLS_DEPLOY_JARS = [
     "//conditions:default": [],
 })
 
-JAVA_VERSIONS = ("11",)
+JAVA_VERSIONS = ("11", "15")
```
Comment (comius) [inline src/BUILD:573]: Remove "15" from here. This is for testing java_tools releases.

File: src/upload_all_java_tools.sh [hunk 5]
```diff
@@ -48,7 +48,7 @@ bazel_version=$(bazel info release | cut -d' ' -f2)
 
 # Passing the same commit_hash and timestamp to all targets to mark all the artifacts
 # uploaded on GCS with the same identifier.
-for java_version in 11; do
+for java_version in 11 15; do
```
Comment (comius) [inline src/upload_all_java_tools.sh:51]: Also from here. This is for releasing it.

File: src/BUILD (hunk 4 above)
Comment (Jonpez2) [reply src/BUILD]: Sorry, done

File: src/upload_all_java_tools.sh (hunk 5 above)
Comment (Jonpez2) [reply src/upload_all_java_tools.sh]: Sorry, done

File: src/BUILD [hunk 6]
```diff
@@ -761,6 +764,9 @@ filegroup(
         "@remote_java_tools_javac11_test_darwin//:WORKSPACE",
         "@remote_java_tools_javac11_test_linux//:WORKSPACE",
         "@remote_java_tools_javac11_test_windows//:WORKSPACE",
+        "@remote_java_tools_javac14_test_darwin//:WORKSPACE",
```
Comment (comius) [inline src/BUILD:767]: There were removed already.

Comment (Jonpez2) [reply src/BUILD]: Sorry again!  Bad merge.  Fixed.

File: src/BUILD (hunk 4 above)
Comment (davido) [reply src/BUILD]: Without this change, I cannot build java_toolchain 15.

I had to apply this patch:

```diff
diff --git a/src/BUILD b/src/BUILD
index 215b656fc8..419a122013 100644
--- a/src/BUILD
+++ b/src/BUILD
@@ -570,7 +570,7 @@ JAVA_TOOLS_DEPLOY_JARS = [
     "//conditions:default": [],
 })
 
-JAVA_VERSIONS = ("11",)
+JAVA_VERSIONS = ("11","15")
 
 [
     genrule(
```

To create `java_toolchain_java15.zip`:

```
  $ bazel build src:java_tools_java15.zip
INFO: Analyzed target //src:java_tools_java15.zip (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //src:java_tools_java15.zip up-to-date:
  bazel-bin/src/java_tools_java15.zip
INFO: Elapsed time: 0.119s, Critical Path: 0.00s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
```

With that java_toolchain_java15.zip I can build and run with Java 15:

```shell
  $ bazel run java:Javac15Example
INFO: Analyzed target //java:Javac15Example (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //java:Javac15Example up-to-date:
  bazel-bin/java/Javac15Example.jar
  bazel-bin/java/Javac15Example
INFO: Elapsed time: 0.090s, Critical Path: 0.01s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
Hello,
JDK 15
```

The source code for that test project is [here](https://github.com/davido/bazel_java_toolchain14).

Comment (comius) [reply src/BUILD]: Sorry for the confusion.
You should create java_toolchain11.zip and with that you should be able to build either Java 11, Java 14 or Java 15.

File: src/BUILD [hunk 7]
```diff
@@ -752,6 +752,9 @@ filegroup(
         "@openjdk14_darwin_archive//:WORKSPACE",
         "@openjdk14_linux_archive//:WORKSPACE",
         "@openjdk14_windows_archive//:WORKSPACE",
+        "@openjdk15_darwin_archive//:WORKSPACE",
```
Comment (comius) [inline src/BUILD:755]: Please add also
```        
        "@remotejdk15_linux_for_testing//:WORKSPACE",
        "@remotejdk15_macos_for_testing//:WORKSPACE",
        "@remotejdk15_win_for_testing//:WORKSPACE",
```

to test_repos filegroup.

Comment (Jonpez2) [reply src/BUILD:755]: Done, thanks!

File: src/test/shell/bazel/BUILD [hunk 8]
```diff
@@ -202,6 +202,30 @@ sh_test(
...
 )
 
+sh_test(
+    name = "bazel_java15_test",
+    srcs = ["bazel_java15_test.sh"],
+    args = [
+        # --java_toolchain and --host_java_toolchain values
+        "@local_java_tools//:toolchain_jdk_15",
+        # java_tools zip to test
+        "src/java_tools_java15.zip",
```
Comment (comius) [inline src/test/shell/bazel/BUILD:212]: `src/java_tools_java11.zip` here

File: src/test/shell/bazel/BUILD [hunk 9]
```diff
@@ -202,6 +202,30 @@ sh_test(
...
+    ] + select({
+        # --javabase and --host_javabase values
+        "//src/conditions:darwin": ["@openjdk15_darwin_archive//:runtime"],
+        "//src/conditions:darwin_x86_64": ["@openjdk15_darwin_archive//:runtime"],
+        "//src/conditions:windows": ["@openjdk15_windows_archive//:runtime"],
+        "//src/conditions:linux_x86_64": ["@openjdk15_linux_archive//:runtime"],
+    }),
+    data = [
+        ":test-deps",
+        "//src:java_tools_java15_zip",
```
Comment (comius) [inline src/test/shell/bazel/BUILD:222]: and `"//src:java_tools_java11_zip"` here...

tools 11 will have java 14 and java 15

**Review by comius — CHANGES_REQUESTED** [review]
After this fix all the tests should pass.

File: src/test/shell/bazel/BUILD (hunk 8 above)
Comment (Jonpez2) [reply src/test/shell/bazel/BUILD]: Nice thank you, done.

File: src/test/shell/bazel/BUILD (hunk 9 above)
Comment (Jonpez2) [reply src/test/shell/bazel/BUILD]: Nice thank you, done.

**Review by comius — APPROVED** [review]
Thanks!
